---
title: "NN Paper Final Script: Seurat preprocessing"
author: "Mariano Ruz Jurado"
date: "2024-05-14"
output: html_document
editor_options: 
  chunk_output_type: console
---

#0. Load essential libraries
```{r}
#load libraries
library(Seurat)
library(tidyverse)
library(svglite)
library(ggplot2)
library(ggpubr)
library(openxlsx)
source("/media/EOS_ZMM_shared/Bioinformatic/scRNA-SEQ-ZMM/Import10X-HelperFunctions_SeuratV3.R")
source("/media/EOS_ZMM_shared/Bioinformatic/scRNA-SEQ-ZMM/Mariano_functions.R")
outputFolder <- "/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code"
sink(file = paste0(outputFolder,"/NN_Paper_Script.log"), append = TRUE, split = TRUE)
```

#1. Preprocessing data (Integration of Human and Mice Data)

##1.1 Import count matrixes from cellranger outputs
```{r}
#Human Hypertrophy HFrEF, HFpEF
  Sample.Paths.human <- c(
    #AS
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/Hypertrophy(AS)/cellranger/103837-001-001/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/Hypertrophy(AS)/cellranger/103837-001-002/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/Hypertrophy(AS)/cellranger/103837-001-003/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/Hypertrophy(AS)/cellranger/103837-001-004/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/Hypertrophy(AS)/cellranger/103799-001-001//outs/filtered_feature_bc_matrix",
    #HFpEF
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/HFpEF/cellranger/104383-009-005/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/HFpEF/cellranger/104383-009-006/outs/filtered_feature_bc_matrix",
    #HFrEF
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/ICM/cellranger/103837-001-005/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/ICM/cellranger/104383-022/outs/filtered_feature_bc_matrix",
    "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Human/ICM/cellranger/104383-009-004/outs/filtered_feature_bc_matrix"
    )
  Samplenames.human <- c("Human-AS-n2","Human-AS-n4","Human-AS-n5","Human-AS-n1","Human-AS-n3"
                         ,"Human-HFpEF-n1","Human-HFpEF-n2"
                         ,"Human-HFrEF-n1","Human-HFrEF-n2","Human-HFrEF-n3"
                         )

 #Mice ALL cellranger
 Sample.Paths.mice <- c(
   #AS
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Hypertrophy(TAC)/cellranger/104383-009-001/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Hypertrophy(TAC)/cellranger/104383-009-002/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Hypertrophy(TAC)/cellranger/104383-009-003/outs/filtered_feature_bc_matrix",
   #HFpEF
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFpEF/cellranger/104081-001-001/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFpEF/cellranger/104081-001-002/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFpEF/cellranger/104081-001-006/outs/filtered_feature_bc_matrix",
   #HFrEF
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFrEF(ICM)/cellranger/104383-013-001/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFrEF(ICM)/cellranger/104383-013-002/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFrEF(ICM)/cellranger/104383-013-003/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/HFrEF(ICM)/cellranger/104383-013-004/outs/filtered_feature_bc_matrix",
   #Ctrl
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/104081-001-003/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/104081-001-004/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/104081-001-005/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/E-MTAB-7869/old1/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/E-MTAB-7869/old2/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/E-MTAB-7869/old3/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/E-MTAB-7869/young1/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/E-MTAB-7869/young2/outs/filtered_feature_bc_matrix",
   "/media/Helios_scStorage/David/HFrEF_HumanMice_Comparisson/Data/Mice/Control/cellranger/E-MTAB-7869/young3/outs/filtered_feature_bc_matrix"
   )

 Samplenames.mice <- c(
                       "Mice-AS-n1","Mice-AS-n2","Mice-AS-n3"
                       ,"Mice-HFpEF-n1","Mice-HFpEF-n2","Mice-HFpEF-n3"
                       ,"Mice-HFrEF-n1","Mice-HFrEF-n2", "Mice-HFrEF-n3", "Mice-HFrEF-n4"
                       ,"Mice-CTRL-n1","Mice-CTRL-n2","Mice-CTRL-n3","Mice-CTRL-n4","Mice-CTRL-n5"
                       ,"Mice-CTRL-n6","Mice-CTRL-n7","Mice-CTRL-n8","Mice-CTRL-n9"
                       )

```

##1.2 Run Importer function on Human Data
```{r}
outputFolder <- "/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Human_Mice_Integrated"
#import data, create SeuratObjects
tmpList<-list()
SeuratObjectList.human <- list()
for (i in 1:length(Sample.Paths.human)) {
  tmpList<-Importer(pathway = Sample.Paths.human[i],id = Samplenames.human[i], FilterCells = TRUE,FilterByAbsoluteValues = TRUE, performScaling = TRUE, minFeatures=300, maxFeatures=6000,minCounts=500,maxCounts=15000, maxMito=0.05)
  print(tmpList[[2]])
  SeuratObjectList.human[[i]]<-tmpList[[1]]
}

#mapping summary
Mapping_Summary.List <- list()
for (i in 1:length(Sample.Paths.human)) {
Mapping_Summary.List[[i]] <- SummarizeMapping(pathway = Sample.Paths.human[i],id = Samplenames.human[i])
}

for (j in 2:length(Mapping_Summary.List)) {
  Mapping_Summary.List[[1]][,j] <- Mapping_Summary.List[[j]]
  mergeDF <- Mapping_Summary.List[[1]]
}

colnames(mergeDF) <- Samplenames.human
write.xlsx(mergeDF, file =paste0(outputFolder,"/Human_mapping_summary.xlsx"),col.names = T,row.names = T)

#clean up
rm(tmpList)
rm(mergeDF)
rm(Mapping_Summary.List)
```

###1.2.1 Load in Human cell atlas Control heart left ventricle Seurat object and Control interventricular septum Seurat object
```{r}
#load LV human control object
SeuratObject.human.combined.LV<- readRDS(file="/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Human_Mice_Integrated/Seurat_and_R_objects/SeuratObject.combined.CTRL.LV.rds")
#load SP human control object
SeuratObject.human.combined.SP<- readRDS(file="/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Human_Mice_Integrated/Seurat_and_R_objects/SeuratObject.combined.CTRL.SP.rds")

#into list
SeuratObjectList.human<-c(SeuratObjectList.human,SeuratObject.human.combined.LV,SeuratObject.human.combined.SP)

#clean up
rm(SeuratObject.human.combined.LV)
rm(SeuratObject.human.combined.SP)

```

##1.3 Run Importer function on Mice Data
```{r}
outputFolder <- "/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Human_Mice_Integrated"
tmpList<-list()
SeuratObjectList.mice <- list()
for (i in 1:length(Sample.Paths.mice)) {
  tmpList<-Importer(pathway = Sample.Paths.mice[i],id = Samplenames.mice[i], FilterCells = TRUE,FilterByAbsoluteValues = TRUE, performScaling = TRUE,minFeatures=300, maxFeatures=6000,minCounts=500,maxCounts=15000, maxMito=0.05)
  print(tmpList[[2]])
  SeuratObjectList.mice[[i]]<-tmpList[[1]]
}

#mapping summary
Mapping_Summary.List <- list()
for (i in 1:length(Sample.Paths.mice)) {
Mapping_Summary.List[[i]] <- SummarizeMapping(pathway = Sample.Paths.mice[i],id = Samplenames.mice[i])
}

for (j in 2:length(Mapping_Summary.List)) {
  Mapping_Summary.List[[1]][,j] <- Mapping_Summary.List[[j]]
  mergeDF <- Mapping_Summary.List[[1]]
}

colnames(mergeDF) <- Samplenames.mice
write.xlsx(mergeDF, file =paste0(outputFolder,"/Mice_mapping_summary.xls"), col.names = T,row.names = T)

#clean up
rm(tmpList)
rm(mergeDF)
rm(Mapping_Summary.List)
```

##1.4 Add disease and species information to the objects before integrating 
```{r}
#add species to human objects
for (i in 1:length(Samplenames.human)) {
  SeuratObjectList.human[[i]]$species <- "Human"
}
SeuratObjectList.human[[1]]$disease<-"Hypertrophy"
SeuratObjectList.human[[2]]$disease<-"Hypertrophy"
SeuratObjectList.human[[3]]$disease<-"Hypertrophy"
SeuratObjectList.human[[4]]$disease<-"Hypertrophy"
SeuratObjectList.human[[5]]$disease<-"Hypertrophy"
SeuratObjectList.human[[6]]$disease<-"HFpEF"
SeuratObjectList.human[[7]]$disease<-"HFpEF"
SeuratObjectList.human[[8]]$disease<-"HFrEF"
SeuratObjectList.human[[9]]$disease<-"HFrEF"
SeuratObjectList.human[[10]]$disease<-"HFrEF"

#add species to mice objects
for (i in 1:length(Samplenames.mice)) {
  SeuratObjectList.mice[[i]]$species <- "Mice"  
}

SeuratObjectList.mice[[1]]$disease<-"Hypertrophy"
SeuratObjectList.mice[[2]]$disease<-"Hypertrophy"
SeuratObjectList.mice[[3]]$disease<-"Hypertrophy"
SeuratObjectList.mice[[4]]$disease<-"HFpEF"
SeuratObjectList.mice[[5]]$disease<-"HFpEF"
SeuratObjectList.mice[[6]]$disease<-"HFpEF"
SeuratObjectList.mice[[7]]$disease<-"HFrEF"
SeuratObjectList.mice[[8]]$disease<-"HFrEF"
SeuratObjectList.mice[[9]]$disease<-"HFrEF"
SeuratObjectList.mice[[10]]$disease<-"HFrEF"
SeuratObjectList.mice[[11]]$disease<-"CTRL"
SeuratObjectList.mice[[12]]$disease<-"CTRL"
SeuratObjectList.mice[[13]]$disease<-"CTRL"
SeuratObjectList.mice[[14]]$disease<-"CTRL"
SeuratObjectList.mice[[15]]$disease<-"CTRL"
SeuratObjectList.mice[[16]]$disease<-"CTRL"
SeuratObjectList.mice[[17]]$disease<-"CTRL"
SeuratObjectList.mice[[18]]$disease<-"CTRL"
SeuratObjectList.mice[[19]]$disease<-"CTRL"

```

## 1.5 Apply thresholds from Human Healthy ATLAS on data
```{r}
#human
for (i in 1:length(SeuratObjectList.human)) {
  Object <- SeuratObjectList.human[[i]]
  Object <- subset(Object, subset = percent.mito < 0.05 &
                                          nFeature_RNA > 300 & nFeature_RNA < 6000 &
                                          nCount_RNA > 500 & nCount_RNA <15000)
  SeuratObjectList.human[[i]] <- Object
}

#mice
for (i in 1:length(SeuratObjectList.mice)) {
  Object <- SeuratObjectList.mice[[i]]
  Object <- subset(Object, subset = percent.mito < 0.05 &
                                          nFeature_RNA > 300 & nFeature_RNA < 6000 &
                                          nCount_RNA > 500 & nCount_RNA <15000)
  SeuratObjectList.mice[[i]] <- Object
}

#clean up
rm(Object)
```

## 1.6 Build dataframe with orthologs across Human and Mice + subset and Integration
```{r}
library(OrthoIntegrate)

# Build Orthologues using OrthoIntegrate algorithm
Ortholog.DF <- BuildOrthologues(GTF.1 = "/opt/refdata-gex-GRCh38-2020-A/genes/genes.gtf",
                                GTF.2 = "/opt/refdata-gex-mm10-2020-A/genes/genes.gtf",
                                species.1 = "human",
                                species.2 = "mouse")

# Integrate across species using only 1 to 1 assigned genes
SeuratObject.combined <- IntegrateObjects(OrthologueList = Ortholog.DF,
                                          SeuratObjectList.species.1 = SeuratObjectList.human,
                                          SeuratObjectList.species.2 = SeuratObjectList.mice,
                                          species.1 = "human",
                                          species.2 = "mouse")

```

##1.7 Scaling, PCA, UMAP, Clustering + Cell type annotation using CellTypist python package

```{r}

DefaultAssay(object = SeuratObject.combined) <- "integrated"
# Run the standard workflow for visualization and clustering
SeuratObject.combined <- ScaleData(object = SeuratObject.combined, verbose = FALSE)
SeuratObject.combined <- RunPCA(object = SeuratObject.combined, npcs = 30, verbose = FALSE)
# UMAP and Clustering
SeuratObject.combined <- RunUMAP(object = SeuratObject.combined, reduction = "pca", dims = 1:10)
SeuratObject.combined <- FindNeighbors(object = SeuratObject.combined, reduction = "pca", dims = 1:10)
SeuratObject.combined <- FindClusters(SeuratObject.combined, resolution = 0.3)


Seu.Obj.annotated <- DO.CellTypist(seuratObj = SeuratObject.combined,
                                                 modelName = "Healthy_Adult_Heart.pkl",
                                                 minCellsToRun = 200,
                                                 runCelltypistUpdate = TRUE,
                                                 over_clustering = "seurat_clusters",
                                                 assay_normalized = "RNA")

p1 <- DimPlot(Seu.Obj.annotated, group.by = "predicted_labels_celltypist", label = T, split.by = "species")
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Annotation_celltypist_split_species.png",plot=p1,width = 11, height = 7)

# Re-annotate
Seu.Obj.annotated$cell_type <- plyr::revalue(Seu.Obj.annotated$predicted_labels_celltypist,
                                             c(`FB1` = "Fibroblasts",
                                               `PC1_vent` = "Pericytes",
                                               `EC1_cap` = "Endothelial",
                                               `NC1_glial` = "Neuro",
                                               `SMC1_basic` = "Smooth.Muscle",
                                               `Meso` = "Pericytes",
                                               `Mast` = "Cardiomyocytes",
                                               `EC5_art` = "Pericytes",
                                               `LYVE1+IGF1+MP` = "Immune.cells",
                                               `vCM1` = "Cardiomyocytes",
                                               `vCM3_stressed` = "Cardiomyocytes"))
#Human
Seu.Obj.annotated.H <- subset(Seu.Obj.annotated, subset = species == "Human")
p1 <- DimPlot(Seu.Obj.annotated.H, group.by = "cell_type", label = F, raster = F)
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Annotation_manually_curated_human.png",plot=p1,width = 15, height = 15)
#Mouse
Seu.Obj.annotated.M <- subset(Seu.Obj.annotated, subset = species == "Mice")
p1 <- DimPlot(Seu.Obj.annotated.M, group.by = "cell_type", label = F, raster = F)
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Annotation_manually_curated_mouse.png",plot=p1,width = 15, height = 15)

# per orig.ident
#UMAPs by species 
p1 <- DimPlot(Seu.Obj.annotated, reduction = "umap", group.by = "species", pt.size = 0.01, label = F, raster = F, cols = c("#1f77b4","#ff7f0e"))
#set alpha values for better visuals
p1[[1]]$layers[[1]]$aes_params$alpha = ifelse ( Seu.Obj.annotated@meta.data$species == "Mice", 1, .4)
p1 <- p1 + theme(legend.position = "none")+
  ggtitle("")+
  xlab("")+
  ylab("")
p1
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/UMAP_per_sample.png",plot=p1,width = 15, height = 15)


#For celltypist annotation
#Human
Seu.Obj.annotated.H <- subset(Seu.Obj.annotated, subset = species == "Human")
p1 <- DimPlot(Seu.Obj.annotated.H, group.by = "predicted_labels_celltypist", label = F, raster = F)
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Annotation_celltypist_human.png",plot=p1,width = 15, height = 15)
#Mouse
Seu.Obj.annotated.M <- subset(Seu.Obj.annotated, subset = species == "Mice")
p1 <- DimPlot(Seu.Obj.annotated.M, group.by = "predicted_labels_celltypist", label = F, raster = F)
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Annotation_celltypist_mouse.png",plot=p1,width = 15, height = 15)

#For unsupervised clustering
#Human
Seu.Obj.annotated.H <- subset(Seu.Obj.annotated, subset = species == "Human")
p1 <- DimPlot(Seu.Obj.annotated.H, group.by = "seurat_clusters", label = F, raster = F)
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Unsupervised_clustering_human.png",plot=p1,width = 15, height = 15)
#Mouse
Seu.Obj.annotated.M <- subset(Seu.Obj.annotated, subset = species == "Mice")
p1 <- DimPlot(Seu.Obj.annotated.M, group.by = "seurat_clusters", label = F, raster = F)
ggsave("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/UMAPs/Unsupervised_clustering_mouse.png",plot=p1,width = 15, height = 15)

#saveRDS(SeuratObject.combined.annotated, file = "/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Human_Mice_Integrated/Seurat_and_R_objects/SeuratObject.combined.integrated.annotated_10_08_22.rds")
```

#2. Quality Control Figure (nCount, nFeatures, percent mitochondrial content per Sample)
```{r}
DefaultAssay(Seu.Obj.annotated) <- "RNA"
Idents(Seu.Obj.annotated) <- "condition"

#some re-assignments due to a typo made by object creation
Seu.Obj.annotated$orig.ident <- plyr::revalue(Seu.Obj.annotated$orig.ident,
                                              c(`Human_CTRLsp-n1` = "Human-CTRLsp-n1",
                                                `Human_CTRLsp-n2` = "Human-CTRLsp-n2",
                                                `Human_CTRLsp-n3` = "Human-CTRLsp-n3",
                                                `Human_CTRLsp-n4` = "Human-CTRLsp-n4",
                                                `Human_CTRLsp-n5` = "Human-CTRLsp-n5",
                                                `Human_CTRLsp-n6` = "Human-CTRLsp-n6",
                                                `Human_CTRLsp-n7` = "Human-CTRLsp-n7",
                                                `Human_CTRLsp-n8` = "Human-CTRLsp-n8",
                                                `Human_CTRLsp-n9` = "Human-CTRLsp-n9",
                                                `Human_CTRLsp-n10` = "Human-CTRLsp-n10",
                                                `Human_CTRLsp-n11` = "Human-CTRLsp-n11",
                                                `Human_CTRLsp-n12` = "Human-CTRLsp-n12"))
table(Seu.Obj.annotated$orig.ident)
#Get  cell number counts for individual classes
Seu.Obj.annotated$condition <- plyr::revalue(Seu.Obj.annotated$condition,
                                                   c(`Human_CTRLsp` = "Human-CTRL",
                                                     `Human-CTRLlv` = "Human-CTRL"))
table(Seu.Obj.annotated$condition)

# make quality plots for human and mice (nFeatures)

for (spc in c("Human", "Mice")) {
  Seu.Obj.annotated.sub <- subset(Seu.Obj.annotated, subset = species == spc)

  df<-data.frame(orig.ident=setNames(Seu.Obj.annotated.sub[["orig.ident"]][,"orig.ident"], rownames(Seu.Obj.annotated.sub[["orig.ident"]])))
  
  #get expression values for genes from individual cells, add to df
  df[,"nFeatures"] <- Seu.Obj.annotated.sub$nFeature_RNA
  df.melt <- melt(df)
  
  ctrl_entries = grep("Human-CTRL", unique(df.melt$orig.ident),value = T)
  rest_entries = unique(df.melt$orig.ident)[!unique(df.melt$orig.ident) %in% ctrl_entries]
  
  df.melt$orig.ident <- factor(df.melt$orig.ident, levels = c(ctrl_entries, rest_entries))
  
  plot_colors <- c(rep("#eeeeee", 53))
  p <- ggplot(df.melt, aes(x = orig.ident, y = value))+ 
    geom_violin(aes(fill = orig.ident), trim = T, scale = "width", size = 1)+
    geom_boxplot(size = 0.40, width = 0.25, alpha = 0.5, color = "darkred", outlier.size = 0)+
    labs(title = "Number of Features per sample", y = "nFeatures")+
    xlab("")+
    theme_light()+
    theme(plot.title = element_text(face = "bold", color = "black", hjust = 0.5, size = 14),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          axis.text.x = element_text(face = "bold", color = "black", angle = 45, hjust = 1, size = 14),
          axis.text.y = element_text(face = "bold", color = "black", hjust = 1, size = 14),
          legend.position = "none")+
    scale_fill_manual(values = plot_colors)
  p
  ggsave(plot = p, filename = paste0("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/quality_vlnplots/nFeature_vlnplot_",spc, ".pdf"), width = 12, height = 7)
}

# make quality plots for human and mice (nCounts)

for (spc in c("Human", "Mice")) {
  Seu.Obj.annotated.sub <- subset(Seu.Obj.annotated, subset = species == spc)

  df<-data.frame(orig.ident=setNames(Seu.Obj.annotated.sub[["orig.ident"]][,"orig.ident"], rownames(Seu.Obj.annotated.sub[["orig.ident"]])))
  
  #get expression values for genes from individual cells, add to df
  df[,"nCount"] <- Seu.Obj.annotated.sub$nCount_RNA
  df.melt <- melt(df)
  
  ctrl_entries = grep("Human-CTRL", unique(df.melt$orig.ident),value = T)
  rest_entries = unique(df.melt$orig.ident)[!unique(df.melt$orig.ident) %in% ctrl_entries]
  
  df.melt$orig.ident <- factor(df.melt$orig.ident, levels = c(ctrl_entries, rest_entries))
  
  plot_colors <- c(rep("#eeeeee", 53))
  p <- ggplot(df.melt, aes(x = orig.ident, y = value))+ 
    geom_violin(aes(fill = orig.ident), trim = T, scale = "width", size = 1)+
    geom_boxplot(size = 0.40, width = 0.25, alpha = 0.5, color = "darkred", outlier.size = 0)+
    labs(title = "Number of Counts per sample", y = "nCounts")+
    xlab("")+
    theme_light()+
    theme(plot.title = element_text(face = "bold", color = "black", hjust = 0.5, size = 14),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          axis.text.x = element_text(face = "bold", color = "black", angle = 45, hjust = 1, size = 14),
          axis.text.y = element_text(face = "bold", color = "black", hjust = 1, size = 14),
          legend.position = "none")+
    scale_fill_manual(values = plot_colors)
  p
  ggsave(plot = p, filename = paste0("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/quality_vlnplots/nCount_vlnplot_",spc, ".pdf"), width = 12, height = 7)
}



# make quality plots for human and mice (nCounts)

for (spc in c("Human", "Mice")) {
  Seu.Obj.annotated.sub <- subset(Seu.Obj.annotated, subset = species == spc)

  df<-data.frame(orig.ident=setNames(Seu.Obj.annotated.sub[["orig.ident"]][,"orig.ident"], rownames(Seu.Obj.annotated.sub[["orig.ident"]])))
  
  #get expression values for genes from individual cells, add to df
  df[,"percentmito"] <- Seu.Obj.annotated.sub$percent.mito
  df.melt <- melt(df)
  
  ctrl_entries = grep("Human-CTRL", unique(df.melt$orig.ident),value = T)
  rest_entries = unique(df.melt$orig.ident)[!unique(df.melt$orig.ident) %in% ctrl_entries]
  
  df.melt$orig.ident <- factor(df.melt$orig.ident, levels = c(ctrl_entries, rest_entries))
  
  plot_colors <- c(rep("#eeeeee", 53))
  p <- ggplot(df.melt, aes(x = orig.ident, y = value))+ 
    geom_violin(aes(fill = orig.ident), trim = T, scale = "width", size = 1)+
    geom_boxplot(size = 0.40, width = 0.25, alpha = 0.5, color = "darkred", outlier.size = 0)+
    labs(title = "Percentage of mitochondrial content per sample", y = "perc.mito")+
    xlab("")+
    theme_light()+
    theme(plot.title = element_text(face = "bold", color = "black", hjust = 0.5, size = 14),
          axis.title.y = element_text(face = "bold", color = "black", size = 14),
          axis.text.x = element_text(face = "bold", color = "black", angle = 45, hjust = 1, size = 14),
          axis.text.y = element_text(face = "bold", color = "black", hjust = 1, size = 14),
          legend.position = "none")+
    scale_fill_manual(values = plot_colors)
  p
  ggsave(plot = p, filename = paste0("/media/Helios_scStorage/Mariano/NN_Human_Mice/Paper_R_code/quality_vlnplots/percentmito_vlnplot_",spc, ".pdf"), width = 12, height = 7)
}

```


#3. Find Markers per cell type

```{r}
# load object
Seu.Obj.annotated <- readRDS(file = "/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Human_Mice_Integrated/Seurat_and_R_objects/SeuratObject.combined.integrated.annotated_10_08_22.rds")
DefaultAssay(object = Seu.Obj.annotated) <- "RNA"

#create species specific sub sets of whole data set
Seu.Obj.annotated.H <- subset(Seu.Obj.annotated, subset = species == "Human")
Idents(Seu.Obj.annotated.H) <-"cell_type"

Seu.Obj.annotated.M <- subset(Seu.Obj.annotated, subset = species == "Mice")
Idents(Seu.Obj.annotated.M) <-"cell_type"

# Run Seurat Marker function over annotation
Obj.Markers.in.cell_type.human <- FindAllMarkers(object = Seu.Obj.annotated.H, min.pct = 0.25, logfc.threshold = 0.25, only.pos = T, test.use = "wilcox")
Obj.Markers.in.cell_type.mice <- FindAllMarkers(object = Seu.Obj.annotated.M, min.pct = 0.25, logfc.threshold = 0.25, only.pos = T, test.use = "wilcox")

# Save as an excel
write.xlsx(Obj.Markers.in.cell_type.human,file=paste0(outputFolder,"/excelsheets/markers_celtype/Human_markers_cell_annotation.xlsx"), colNames = T, rowNames = T)
write.xlsx(Obj.Markers.in.cell_type.mice,file=paste0(outputFolder,"/excelsheets/markers_celtype/Mouse_markers_cell_annotation.xlsx"), colNames = T, rowNames = T)

top5<-Obj.Markers.in.cell_type.human %>% dplyr::group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
#kick MT-genes
MT.genes <- top5[grep(pattern = "MT-", top5$gene),]
tp5 <- top5[!(top5$gene %in% MT.genes$gene),]
write.xlsx(tp5,file=paste0(outputFolder,"/excelsheets/markers_celtype/Human_tp5_markers_cell_annotation.xlsx"), colNames = T, rowNames = T)

top5<-Obj.Markers.in.cell_type.mice %>% dplyr::group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
#kick MT-genes
MT.genes <- top5[grep(pattern = "MT-", top5$gene),]
tp5 <- top5[!(top5$gene %in% MT.genes$gene),]
write.xlsx(tp5,file=paste0(outputFolder,"/excelsheets/markers_celtype/Mouse_tp5_markers_cell_annotation.xlsx"), colNames = T, rowNames = T)

# Personally preference for scanpy's dotplot therefore convert to anndata for it and continue in scanpy_dotplot.py
sce.Obj.annotated.H <- Seurat::as.SingleCellExperiment(Seu.Obj.annotated.H)
zellkonverter::writeH5AD(sce.Obj.annotated.H, file = paste0(outputFolder,"/objects/adata.Obj.annotated.H.h5ad"), X_name = "logcounts")

sce.Obj.annotated.M <- Seurat::as.SingleCellExperiment(Seu.Obj.annotated.M)
zellkonverter::writeH5AD(sce.Obj.annotated.M, file = paste0(outputFolder,"/objects/adata.Obj.annotated.M.h5ad"), X_name = "logcounts")

```

#4. Extracting normalized expression data for machine learning training for train, val data

```{r}
Idents(Seu.Obj.annotated) <- "orig.ident"
orig.ident <- levels(Idents(Seu.Obj.annotated))
#get the subset obj
Seu.Obj <- subset(Seu.Obj.annotated, idents = orig.ident[!orig.ident %in% c("Mice-AS-n2",
                                                                  "Mice-HFpEF-n1",
                                                                  "Mice-HFrEF-n2",
                                                                  "Mice-CTRL-n6",
                                                                  "Human-AS-n4",
                                                                  "Human-HFpEF-n2",
                                                                  "Human-HFrEF-n2",
                                                                  "Human-CTRLlv-n2",
                                                                  "Human_CTRLsp-n2")])
#rename every Human-CTRL(sp,lv) to Human-CTRL
Idents(Seu.Obj) <- "condition"
Seu.Obj$condition <- plyr::revalue(Seu.Obj$condition,
                                                   c(`Human_CTRLsp` = "Human-CTRL",
                                                     `Human-CTRLlv` = "Human-CTRL"))

#remove AS Neuros from mice, there are too few cells to train for that specific type
remove.barcodes <- rownames(Seu.Obj@meta.data[Seu.Obj@meta.data$condition == "Mice-AS" & Seu.Obj@meta.data$cell_type == "Neuro",])
Seu.Obj <- subset(Seu.Obj, cells = remove.barcodes, invert = TRUE)

meta.data <- Seu.Obj@meta.data
sparse.matrix <- Seu.Obj@assays$RNA@data
dense.matrix <- as.matrix(sparse.matrix)
dense.matrix <- t(dense.matrix)


#Create stratified Training and Testing set
library(caret)

label_list <- paste(Seu.Obj@meta.data$condition,
                    Seu.Obj@meta.data$cell_type, sep = "_")



# Create a stratified partition of the data
set.seed(123)  # for reproducibility
trainIndex <- createDataPartition(label_list, p = 0.80, list = FALSE)

# Split the data into training and test sets
train_set <- dense.matrix[trainIndex,]
val_set <- dense.matrix[-trainIndex,]

# Save the training set
write.table(as.data.frame(train_set), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/train_set_without_n2_samples_and_Mice_AS_Neuro_p_80.csv", sep = ",", col.names = T, row.names = T)


#Save test matrix Matrix
write.table(as.data.frame(val_set), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/val_set_without_n2_samples_and_Mice_AS_Neuro_p_20.csv", sep = ",", col.names = T, row.names = T)



# create a list of labels for training data, save as csv for python import

label_list.train <- character()
pb = txtProgressBar(min = 0, max = length(rownames(train_set)), initial = 0)
for (i in 1:length(rownames(train_set))){
    setTxtProgressBar(pb, i)
    barcode <- rownames(train_set)[i]
    condition <- Seu.Obj@meta.data[barcode,]$condition
    cell_type <- Seu.Obj@meta.data[barcode,]$cell_type
    label_list.train[i] <- paste0(condition,"_",cell_type)
}
close(pb)

write.csv(label_list.train, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/train_set_without_n2_samples_and_Mice_AS_Neuro_labels_p_80.csv", row.names = FALSE)

# Write Ground Truth Training barcode, label
df.train <- data.frame(barcodes=rownames(train_set),label_list.train)
openxlsx::write.xlsx(df.train, file = "/media/Helios_scStorage/Mariano/NN_Human_Mice/label_distribution/Ground_Truth_DF_Training.xlsx")

#create a list of labels for my training data, save as csv for python import

label_list.val <- character()
pb = txtProgressBar(min = 0, max = length(rownames(val_set)), initial = 0)
for (i in 1:length(rownames(val_set))){
    setTxtProgressBar(pb, i)
    barcode <- rownames(val_set)[i]
    condition <- Seu.Obj@meta.data[barcode,]$condition
    cell_type <- Seu.Obj@meta.data[barcode,]$cell_type
    label_list.val[i] <- paste0(condition,"_",cell_type)
}
close(pb)

write.csv(label_list.val, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/val_set_without_n2_samples_and_Mice_AS_Neuro_labels_p_20.csv", row.names = FALSE)

#Write Ground Truth val barcode, label
df.val <- data.frame(barcodes=rownames(val_set),label_list.val)
openxlsx::write.xlsx(df.val, file = "/media/Helios_scStorage/Mariano/NN_Human_Mice/label_distribution/Ground_Truth_DF_Val.xlsx")

```

## 4.1 Shuffle randomly Data tables (Good practice for better training)

```{r}
#Load the data sets
train_set <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/train_set_without_n2_samples_and_Mice_AS_Neuro_p_80.csv", header= TRUE, sep = ",")
val_set <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/val_set_without_n2_samples_and_Mice_AS_Neuro_p_20.csv", header= TRUE, sep = ",")

#Load the labels
train_labels <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/train_set_without_n2_samples_and_Mice_AS_Neuro_labels_p_80.csv", sep = ",")
val_labels <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/val_set_without_n2_samples_and_Mice_AS_Neuro_labels_p_20.csv", sep = ",")


#Create randomness in data
set.seed(123) # reproducibility
shuffled_numbers <- sample(nrow(train_set))

#shuffle the frame and labels
shuffled_train_set <- train_set[shuffled_numbers,]
shuffled_train_labels <- as.data.frame(train_labels[shuffled_numbers,])

#check if labels are still correctly assigned, lets check 6 random rows
random_rows <- c(rownames(shuffled_train_set)[1],
  rownames(shuffled_train_set)[11],
  rownames(shuffled_train_set)[111],
  rownames(shuffled_train_set)[1111],
  rownames(shuffled_train_set)[11111],
  rownames(shuffled_train_set)[111111])

#Check if label correct
for (rwnm in random_rows) {
  nbr <- which(rownames(shuffled_train_set) == rwnm)
  label_name <- shuffled_train_labels[nbr,]
  nbr_truth <- which(rownames(train_set) == rwnm)
  truth_label_name <- train_labels[nbr_truth,]
  
  if (label_name == truth_label_name) {
    print(paste0(label_name, " is identical to ", truth_label_name, ", labels identical for the selected random row"))
  }
  
  if (label_name != truth_label_name) {
    print(paste0(label_name, " is NOT identical to ", truth_label_name, ", labels NOT identical for the selected random row"))
  }
}

#if everything is fine save as randomized sample set
write.table(shuffled_train_set, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/RANDOMIZED_train_set_without_n2_samples_and_Mice_AS_Neuro_p_80.csv", sep = ",", col.names = T, row.names = T)

write.csv(shuffled_train_labels, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/RANDOMIZED_train_set_without_n2_samples_and_Mice_AS_Neuro_labels_p_80.csv", row.names = FALSE)


#Now repeat for validation data set

#Create randomness in data
set.seed(123) # reproducibility
shuffled_numbers <- sample(nrow(val_set))

#shuffle the frame and labels
shuffled_val_set <- val_set[shuffled_numbers,]
shuffled_val_labels <- as.data.frame(val_labels[shuffled_numbers,])

#check if labels are still correctly assigned, lets check 6 random rows
random_rows <- c(rownames(shuffled_val_set)[1],
  rownames(shuffled_val_set)[11],
  rownames(shuffled_val_set)[111],
  rownames(shuffled_val_set)[1111],
  rownames(shuffled_val_set)[11111],
  rownames(shuffled_val_set)[44972])

#Check if label correct
for (rwnm in random_rows) {
  nbr <- which(rownames(shuffled_val_set) == rwnm)
  label_name <- shuffled_val_labels[nbr,]
  nbr_truth <- which(rownames(val_set) == rwnm)
  truth_label_name <- val_labels[nbr_truth,]
  
  if (label_name == truth_label_name) {
    print(paste0(label_name, " is identical to ", truth_label_name, ", labels identical for the selected random row"))
  }
  
  if (label_name != truth_label_name) {
    print(paste0(label_name, " is NOT identical to ", truth_label_name, ", labels NOT identical for the selected random row"))
  }
}

#if everything is fine save as randomized sample set
write.table(shuffled_val_set, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/RANDOMIZED_val_set_without_n2_samples_and_Mice_AS_Neuro_p_20.csv", sep = ",", col.names = T, row.names = T)

write.csv(shuffled_val_labels, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/RANDOMIZED_val_set_without_n2_samples_and_Mice_AS_Neuro_labels_p_20.csv", row.names = FALSE)
```

##4.2 Extracting normalized expression data for machine learning training for test data

```{r}

### Create set for testing, best quality samples
Idents(Seu.Obj.annotated) <- "orig.ident"
orig.ident <- levels(Idents(Seu.Obj.annotated))
Seu.Obj.test <- subset(Seu.Obj.annotated, idents = orig.ident[orig.ident %in% c("Mice-AS-n2",
                                                                  "Mice-HFpEF-n1",  
                                                                  "Mice-HFrEF-n2",
                                                                  "Mice-CTRL-n6",
                                                                  "Human-AS-n4", 
                                                                  "Human-HFpEF-n2",
                                                                  "Human-HFrEF-n2",
                                                                  "Human-CTRLlv-n2",
                                                                  "Human_CTRLsp-n2")])

#rename every Human-CTRL(sp,lv) Human-CTRL
Idents(Seu.Obj.test) <- "condition"
Seu.Obj.test$condition <- plyr::revalue(Seu.Obj.test$condition,c(`Human_CTRLsp` = "Human-CTRL",
                                   `Human-CTRLlv` = "Human-CTRL"))

#remove AS Neuros from mice, there are too less cells to train
remove.barcodes <- rownames(Seu.Obj.test@meta.data[Seu.Obj.test@meta.data$condition == "Mice-AS" & Seu.Obj.test@meta.data$cell_type == "Neuro",])
Seu.Obj.test <- subset(Seu.Obj.test, cells = remove.barcodes, invert = TRUE)

meta.data <- Seu.Obj.test@meta.data
sparse.matrix <- Seu.Obj.test@assays$RNA@data
dense.matrix <- as.matrix(sparse.matrix)
dense.matrix <- t(dense.matrix)

label_list <- paste(Seu.Obj.test@meta.data$condition, Seu.Obj.test@meta.data$cell_type, sep = "_")

# Save the test set
write.table(as.data.frame(dense.matrix), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/240519_test_set.csv", sep = ",", col.names = T, row.names = T)

write.csv(label_list, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/240519_test_set_labels_labels.csv", row.names = FALSE)
```

##4.3 Extracting normalized expression data for SHAP value calculation (use HFpEF n1, n2 too small...)

```{r}
Idents(Seu.Obj.annotated) <- "orig.ident"
orig.ident <- levels(Idents(Seu.Obj.annotated))
Seu.Obj.SHAP <- subset(Seu.Obj.annotated, idents = orig.ident[orig.ident %in% c("Mice-AS-n2",
                                                                  "Mice-HFpEF-n1",  
                                                                  "Mice-HFrEF-n2",
                                                                  "Mice-CTRL-n6",
                                                                  "Human-AS-n4", 
                                                                  "Human-HFpEF-n1",
                                                                  "Human-HFrEF-n2",
                                                                  "Human-CTRLlv-n2",
                                                                  "Human_CTRLsp-n2")])

#rename every Human-CTRL(sp,lv) Human-CTRL
Idents(Seu.Obj.SHAP) <- "condition"
Seu.Obj.SHAP$condition <- plyr::revalue(Seu.Obj.SHAP$condition,c(`Human_CTRLsp` = "Human-CTRL",
                                   `Human-CTRLlv` = "Human-CTRL"))

#remove AS Neuros from mice, there are too less cells to train
remove.barcodes <- rownames(Seu.Obj.SHAP@meta.data[Seu.Obj.SHAP@meta.data$condition == "Mice-AS" & Seu.Obj.SHAP@meta.data$cell_type == "Neuro",])
Seu.Obj.SHAP <- subset(Seu.Obj.SHAP, cells = remove.barcodes, invert = TRUE)

meta.data <- Seu.Obj.SHAP@meta.data
sparse.matrix <- Seu.Obj.SHAP@assays$RNA@data
dense.matrix <- as.matrix(sparse.matrix)
dense.matrix <- t(dense.matrix)

label_list <- paste(Seu.Obj.SHAP@meta.data$condition, Seu.Obj.SHAP@meta.data$cell_type, sep = "_")

# Save the test set
write.table(as.data.frame(dense.matrix), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/240716_SHAP_set.csv", sep = ",", col.names = T, row.names = T)

write.csv(label_list, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/240716_SHAP_set_labels_labels.csv", row.names = FALSE)

```


#5. Extracting Human only expression matrix for re-training model on human data only

```{r}
Seu.Obj.annotated.H <- readRDS(file = "/media/Helios_scStorage/Mariano/Human_Mice_Comparison/Seurat_and_R_objects/SeuratObject_Human_AS_HFpEF_HFrEF_Ctrl_Annotated_20_10_22.rds")
DefaultAssay(object = Seu.Obj.annotated.H) <- "RNA"

Idents(Seu.Obj.annotated.H) <- "orig.ident"
orig.ident <- levels(Idents(Seu.Obj.annotated.H))
#get the subset obj
Seu.Obj.H <- subset(Seu.Obj.annotated.H, idents = orig.ident[!orig.ident %in% c("Human-AS-n4",
                                                                  "Human-HFpEF-n2",
                                                                  "Human-HFrEF-n2",
                                                                  "Human-CTRLlv-n2",
                                                                  "Human_CTRLsp-n2")])
#rename every Human-CTRL(sp,lv) to Human-CTRL
Idents(Seu.Obj.H) <- "condition"
Seu.Obj.H$condition <- plyr::revalue(Seu.Obj.H$condition,
                                                   c(`Human_CTRLsp` = "Human-CTRL",
                                                     `Human-CTRLlv` = "Human-CTRL"))


meta.data <- Seu.Obj.H@meta.data
sparse.matrix <- Seu.Obj.H@assays$RNA@data
dense.matrix <- as.matrix(sparse.matrix)
dense.matrix <- t(dense.matrix)


#Create stratified Training and Testing set
library(caret)

label_list <- paste(Seu.Obj.H@meta.data$condition,
                    Seu.Obj.H@meta.data$cell_type, sep = "_")



# Create a stratified partition of the data
set.seed(123)  # for reproducibility
trainIndex <- createDataPartition(label_list, p = 0.80, list = FALSE)

# Split the data into training and test sets
train_set <- dense.matrix[trainIndex,]
val_set <- dense.matrix[-trainIndex,]

# Save the training set
write.table(as.data.frame(train_set), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_train_set_p_80.csv", sep = ",", col.names = T, row.names = T)


#Save val matrix Matrix
write.table(as.data.frame(val_set), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_val_set_p_20.csv", sep = ",", col.names = T, row.names = T)



# create a list of labels for training data, save as csv for python import

label_list.train <- character()
pb = txtProgressBar(min = 0, max = length(rownames(train_set)), initial = 0)
for (i in 1:length(rownames(train_set))){
    setTxtProgressBar(pb, i)
    barcode <- rownames(train_set)[i]
    condition <- Seu.Obj.H@meta.data[barcode,]$condition
    cell_type <- Seu.Obj.H@meta.data[barcode,]$cell_type
    label_list.train[i] <- paste0(condition,"_",cell_type)
}
close(pb)

write.csv(label_list.train, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_train_set_labels_p_80.csv", row.names = FALSE)

# Write Ground Truth Training barcode, label
df.train <- data.frame(barcodes=rownames(train_set),label_list.train)
openxlsx::write.xlsx(df.train, file = "/media/Helios_scStorage/Mariano/NN_Human_Mice/label_distribution/Human_Ground_Truth_DF_Training.xlsx")

#create a list of labels for my training data, save as csv for python import

label_list.val <- character()
pb = txtProgressBar(min = 0, max = length(rownames(val_set)), initial = 0)
for (i in 1:length(rownames(val_set))){
    setTxtProgressBar(pb, i)
    barcode <- rownames(val_set)[i]
    condition <- Seu.Obj.H@meta.data[barcode,]$condition
    cell_type <- Seu.Obj.H@meta.data[barcode,]$cell_type
    label_list.val[i] <- paste0(condition,"_",cell_type)
}
close(pb)

write.csv(label_list.val, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_val_set_labels_p_20.csv", row.names = FALSE)

#Write Ground Truth val barcode, label
df.val <- data.frame(barcodes=rownames(val_set),label_list.val)
openxlsx::write.xlsx(df.val, file = "/media/Helios_scStorage/Mariano/NN_Human_Mice/label_distribution/Human_Ground_Truth_DF_Val.xlsx")

```

## 5.1 Shuffle Human Data tables (Good practice for better training)

```{r}
#Load the data sets
train_set <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_train_set_p_80.csv", header= TRUE, sep = ",")
val_set <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_val_set_p_20.csv", header= TRUE, sep = ",")

#Load the labels
train_labels <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_train_set_labels_p_80.csv", sep = ",")
val_labels <- read.delim(file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_val_set_labels_p_20.csv", sep = ",")


#Create randomness in data
set.seed(123) # reproducibility
shuffled_numbers <- sample(nrow(train_set))

#shuffle the frame and labels
shuffled_train_set <- train_set[shuffled_numbers,]
shuffled_train_labels <- as.data.frame(train_labels[shuffled_numbers,])

#check if labels are still correctly assigned, lets check 6 random rows
random_rows <- c(rownames(shuffled_train_set)[1],
  rownames(shuffled_train_set)[11],
  rownames(shuffled_train_set)[111],
  rownames(shuffled_train_set)[1111],
  rownames(shuffled_train_set)[11111],
  rownames(shuffled_train_set)[111111])

#Check if label correct
for (rwnm in random_rows) {
  nbr <- which(rownames(shuffled_train_set) == rwnm)
  label_name <- shuffled_train_labels[nbr,]
  nbr_truth <- which(rownames(train_set) == rwnm)
  truth_label_name <- train_labels[nbr_truth,]
  
  if (label_name == truth_label_name) {
    print(paste0(label_name, " is identical to ", truth_label_name, ", labels identical for the selected random row"))
  }
  
  if (label_name != truth_label_name) {
    print(paste0(label_name, " is NOT identical to ", truth_label_name, ", labels NOT identical for the selected random row"))
  }
}

#if everything is fine save as randomized sample set
write.table(shuffled_train_set, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/RANDOMIZED_HUMAN_train_set_p_80.csv", sep = ",", col.names = T, row.names = T)

write.csv(shuffled_train_labels, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/RANDOMIZED_HUMAN_train_set_labels_p_80.csv", row.names = FALSE)


#Now repeat for validation data set

#Create randomness in data
set.seed(123) # reproducibility
shuffled_numbers <- sample(nrow(val_set))

#shuffle the frame and labels
shuffled_val_set <- val_set[shuffled_numbers,]
shuffled_val_labels <- as.data.frame(val_labels[shuffled_numbers,])

#check if labels are still correctly assigned, lets check 6 random rows
random_rows <- c(rownames(shuffled_val_set)[1],
  rownames(shuffled_val_set)[11],
  rownames(shuffled_val_set)[111],
  rownames(shuffled_val_set)[1111],
  rownames(shuffled_val_set)[11111],
  rownames(shuffled_val_set)[36135])

#Check if label correct
for (rwnm in random_rows) {
  nbr <- which(rownames(shuffled_val_set) == rwnm)
  label_name <- shuffled_val_labels[nbr,]
  nbr_truth <- which(rownames(val_set) == rwnm)
  truth_label_name <- val_labels[nbr_truth,]
  
  if (label_name == truth_label_name) {
    print(paste0(label_name, " is identical to ", truth_label_name, ", labels identical for the selected random row"))
  }
  
  if (label_name != truth_label_name) {
    print(paste0(label_name, " is NOT identical to ", truth_label_name, ", labels NOT identical for the selected random row"))
  }
}

#if everything is fine save as randomized sample set
write.table(shuffled_val_set, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/RANDOMIZED_HUMAN_val_set_p_20.csv", sep = ",", col.names = T, row.names = T)

write.csv(shuffled_val_labels, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/RANDOMIZED_HUMAN_val_set_labels_p_20.csv", row.names = FALSE)
```

##5.2 Extracting normalized expression data for machine learning training for HUMAN test data
```{r}
Idents(Seu.Obj.annotated.H) <- "orig.ident"
orig.ident <- levels(Idents(Seu.Obj.annotated.H))
#get the subset obj
Seu.Obj.H <- subset(Seu.Obj.annotated.H, idents = orig.ident[orig.ident %in% c("Human-AS-n4",
                                                                  "Human-HFpEF-n2",
                                                                  "Human-HFrEF-n2",
                                                                  "Human-CTRLlv-n2",
                                                                  "Human_CTRLsp-n2")])
#rename every Human-CTRL(sp,lv) to Human-CTRL
Idents(Seu.Obj.H) <- "condition"
Seu.Obj.H$condition <- plyr::revalue(Seu.Obj.H$condition,
                                                   c(`Human_CTRLsp` = "Human-CTRL",
                                                     `Human-CTRLlv` = "Human-CTRL"))


meta.data <- Seu.Obj.H@meta.data
sparse.matrix <- Seu.Obj.H@assays$RNA@data
dense.matrix <- as.matrix(sparse.matrix)
dense.matrix <- t(dense.matrix)

label_list <- paste(Seu.Obj.H@meta.data$condition, Seu.Obj.H@meta.data$cell_type, sep = "_")

# Save the test set
write.table(as.data.frame(dense.matrix), file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_test_set.csv", sep = ",", col.names = T, row.names = T)

write.csv(label_list, file = "/home/mariano/PycharmProjects/MLPGeneExpression/whole_matrix_Seurat_extract/Human_only/Human_test_set_labels.csv", row.names = FALSE)
```

